codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -121,35 +124,91 @@ public void close() { }
     @Override
     public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf data,
             int padding, boolean endStream, ChannelPromise promise) {
-        boolean releaseData = true;
+        boolean lastHeaderWritten = false;
+        boolean lastFrameWritten = false;
         SimpleChannelPromiseAggregator promiseAggregator =
                 new SimpleChannelPromiseAggregator(promise, ctx.channel(), ctx.executor());
+        ByteBuf headerBuf = null;
         try {
             verifyStreamId(streamId, STREAM_ID);
             verifyPadding(padding);
 
-            Http2Flags flags = new Http2Flags().paddingPresent(padding > 0).endOfStream(endStream);
+            // Determine the size of the padding presence field, if padding is present.
+            final int paddingPresenceFieldLength = padding > 0 ? 1 : 0;
 
-            int payloadLength = data.readableBytes() + padding + flags.getPaddingPresenceFieldLength();
-            verifyPayloadLength(payloadLength);
+            int remainingData = data.readableBytes();
 
-            ByteBuf buf = ctx.alloc().buffer(DATA_FRAME_HEADER_LENGTH);
-            writeFrameHeaderInternal(buf, payloadLength, DATA, flags, streamId);
-            writePaddingLength(buf, padding);
-            ctx.write(buf, promiseAggregator.newPromise());
+            // Determine the number of frames that we're about to send. Padding will be split
+            // evenly across the frames, so we subtract off the size of the padding presence field
+            // from each frame (if padding is present).
+            final int numFrames = fastCeil((remainingData + padding) /
+                    (double) (maxFrameSize - paddingPresenceFieldLength));
 
-            // Write the data.
-            releaseData = false;
-            ctx.write(data, promiseAggregator.newPromise());
+            // Split the padding across the frames.
+            final int maxPaddingPerFrame = max(1, fastCeil(padding / (double) numFrames));
 
-            if (padding > 0) { // Write the required padding.
-                ctx.write(ZERO_BUFFER.slice(0, padding).retain(), promiseAggregator.newPromise());
+            // Allocate a buffer to hold the all of the frame headers.
+            headerBuf = ctx.alloc().buffer((numFrames + 1) * DATA_FRAME_HEADER_LENGTH);
```

Comment:
In the case where padding is uniform for every frame (see above note) then if you are writing N frames, then N-1 will have the same header and so you just allocate and fill it out as a buffer and write it N-1 times. The last frame will probably have a different length than the first N-1 (though it might not) and so you write a different buffer for it.